### ISSUE TITLE
Correlated subqueries

### ISSUE THREAD

**Issue by samz406** [ClickHouse/ClickHouse#6697]
I want use the follow sql query, but it does work, in MySQL it is ok.

```sql
SELECT 
(SELECT A1.QTY_SOLD  FROM V_LB_DC_ITEM_CUST_MONTH A1 WHERE A1.CUST_ID=A.CUST_ID LIMIT 1) AS TQ_QTY_SOLD,
A.QTY_SOLD
FROM V_LB_DC_ITEM_CUST_MONTH A
```

See https://materialize.com/blog/?search=Technical+Article how to implement correlated subqueries.

![image](https://user-images.githubusercontent.com/3055810/63832906-589cc780-c9a4-11e9-9c89-d354c1435b3c.png)

**Comment by filimonov** [comment]
Correlated subqueries are not supported currently, and most probably will not be for now. 

You can rewrite your query using JOIN. 

P.S. We still have no issue to track requests for that feature. Let's use this one.

**Comment by leoones** [comment]
> Correlated subqueries are not supported, and currently no schedule for implementing that.
> 
> You can rewrite your query using JOIN.
> 
> P.S. We still have no issue to track requests for that feature. Let's use this one.

Eager to have these functions  ，

**Comment by vadimtk** [comment]
Not being able to use depended subqueries is quite limiting for analytics queries.
I am looking to use  WHERE col1 NOT IN (depended subquery). Not everything can be rewritten using JOIN, especially as ClickHouse LEFT JOIN does not support  inequalities in joins: "JOIN ON inequalities are not supported"

**Comment by kecoliva** [comment]
> Correlated subqueries are not supported, and currently no schedule for implementing that.
> 
> You can rewrite your query using JOIN.
> 
> P.S. We still have no issue to track requests for that feature. Let's use this one.

Would you be able to show a solution using JOIN based from the @samz406's original query?

**Comment by SharonCastel** [comment]
The same issue occurs with window function as it happened with the subquery.
The window function works fine as is, but when putting it inside a case clause Im getting the same error.
BTW I checked the solution also in 2 steps with CTE as the window function and put the identifier inside the case and still the same error occurred.
Is there any plan to fix this issue?

**Comment by den-crane** [comment]
@SharonCastel read this https://github.com/ClickHouse/ClickHouse/issues/19857

**Comment by kahveciderin** [comment]
any update on this?

**Comment by ganeshbawari** [comment]
Hello, Any update in which release this feature will be available? 
Thanks.

**Comment by pmdyy** [comment]
> It is planned for 2022.

Almost 2023. Is it coming anytime soon?

**Comment by alexey-milovidov** [comment]
@pmdyy This task is not going to be solved in 2022.
The current schedule is 2023.

**Comment by lemonwx** [comment]
is there any progress please

**Comment by alexey-milovidov** [comment]
No updates.

**Comment by ArmyDev** [comment]
Is there any news?

**Comment by alexey-milovidov** [comment]
No updates so far.

**Comment by yusufozturk** [comment]
@vdimir, this is one of the issues we discussed at the Amsterdam Meetup. Thanks for help!

**Comment by vdimir** [comment]
@yusufozturk, thank you for pointing that out!

I see that this feature is not currently supported in both the old and new analyzers (`allow_experimental_analyzer = 1`). 

It's a fundamental feature and will require some development effort. However, given its relevance to SQL compatibility, we are encouraged to implement it.

I've discussed this with @KochetovNicolai and @kitaisreal, and we have decided to start working on it after the new analyzer is ready (tracking issue: https://github.com/ClickHouse/ClickHouse/issues/42648).

There are a few possible approaches we can take:
1. Implement it only for limited cases where we can rewrite the query without a correlated subquery (explicit JOIN).
2. Initially, implement a straightforward way that executes the subquery for each row. It may be slower, but it will at least work for small datasets. Then, on top of that, we can start implementing optimizations for specific cases.

So, we are pursuing this feature, but it's important to note that it won't be a quick task to complete.

**Comment by ganesh-ncompass** [comment]
@vdimir - Is this feature done ?

**Comment by alexey-milovidov** [comment]
@ganesh-ncompass, no, and there is not support for correlated subqueries in ClickHouse as of 2024-07-10.

**Comment by ganesh-ncompass** [comment]
@alexey-milovidov @vdimir 

Below is my query. Which isn't right. 
the input_dates has all list of dates against a scheme_code. now i want to get date and respective nav value from nav_details table if that date is present else return most recent previous date's nav.

So I wanted help in performing this using join queries. I tried all possible joins. need solution or some help please.

`dates_query = f'''
WITH input_dates AS (
        SELECT 
            arrayJoin(
                arrayMap(
                    x -> addMonths(start_date, x * {frequency}), 
                    range(toInt32((toRelativeMonthNum(toDate('{end_date}')) - toRelativeMonthNum(toDate(start_date))) / {frequency}) + 1)
                )
            ) AS date,
            scheme_code
        FROM
        (
            SELECT
                scheme_code,
                max(date) AS prev_date,
                argMax(nav, date) AS prev_nav,
                toDate('{start_date}') as start_date
            FROM
                nav_details_temp
            WHERE
                date < '{start_date}' 
                AND scheme_code IN {amfi_scheme_codes}
            GROUP BY
                scheme_code
        ) subquery
)   
SELECT 
    scheme_code,
    date,
    nav
FROM 
    nav_detsils
WHERE
    date <= input_dates.date and scheme_code = input_dates.scheme_code

'''`

**Comment by ganesh-ncompass** [comment]
Also when is the plan to complete this feature, as its already 2024? @alexey-milovidov

**Comment by alexey-milovidov** [comment]
Yes, it is included in the roadmap 2024: https://github.com/ClickHouse/ClickHouse/issues/58392

**Comment by knabbels** [comment]
2024 is over soon, new analyzer is done, any updates here? Please also consider lateral join support.

**Comment by SteZe85** [comment]
The issue mentions a correlated subquery in select statement, which still isn't supported afaik. Is it planned tho?

**Comment by novikd** [comment]
@SteZe85 Yes, there are plans for further correlated subqueries support improvements. Corresponding umbrella issue: https://github.com/ClickHouse/ClickHouse/issues/79890
